### PR TITLE
Add timer

### PR DESCRIPTION
**Pull request by mattkrick**
Trying to stick to ~50 file chunks for easier code review. 
This does NOT include
- styles
- notifications
- timer for any round except the reflect round

TEST
- [ ] only the facilitator can see the timer icon
- [ ] Clicking the timer icon opens up a timer/timebox modal

ASYNC
- [ ] if no one else is in the meeting, the timerbox tab is selected
- [ ] if the team has run a meeting that had the current stage last more than 1 hour, the end time/date is suggested based off of empirical data
- [ ] if the team hasn't used this stage as an async meeting, the end time/date is set to default (~24 hours from now)
- [ ] if a timebox is already active, the update timebox modal opens instead, allowing you to stop or update
- [ ] a countdown exists at the top of the screen when timebox is active

SYNC
- [ ] If someone else is in the meeting, the timer tab is selected by default
- [ ] if the team has run a sync meeting before, the time limit is populated based on empirical data
- [ ] else, the default time limit for the stage is: Reflect- 5 mins, Group - 5 mins, Vote: 2 mins, Discuss - 20 mins (NOTE: reflect phase is currently the only phase wired up!)
- [ ] When the timer is started, a countdown appears for all participants
- [ ] Click the timer icon, then click Stop. The countdown goes away
- [ ] Click the timer icon, add another timer, it works
- [ ] A participant with a clock that is 2 minutes fast still shows the accurate countdown
- [ ] A participant with a clock that is 2 minutes slow still shows the accurate countdown
- [ ] Can click the timer icon, click add time, it works

### CODE REVIEW

**Comment by jordanh** [comment]
# Test Cycle 1

## Tests
- [ ] only the facilitator can see the timer icon
- [ ] Clicking the timer icon opens up a timer/timebox modal

### ASYNC
- [ ] if no one else is in the meeting, the timerbox tab is selected
- [ ] if the team has run a meeting that had the current stage last more than 1 hour, the end time/date is suggested based off of empirical data
- [ ] if the team hasn't used this stage as an async meeting, the end time/date is set to default (~24 hours from now)
- [ ] if a timebox is already active, the update timebox modal opens instead, allowing you to stop or update
- [ ] a countdown exists at the top of the screen when timebox is active

### SYNC
- [ ] If someone else is in the meeting, the timer tab is selected by default
- [ ] if the team has run a sync meeting before, the time limit is populated based on empirical data
- [ ] else, the default time limit for the stage is: Reflect- 5 mins, Group - 5 mins, Vote: 2 mins, Discuss - 20 mins (NOTE: reflect phase is currently the only phase wired up!)
- [ ] When the timer is started, a countdown appears for all participants
- [ ] Click the timer icon, then click Stop. The countdown goes away
- [ ] Click the timer icon, add another timer, it works
- [ ] A participant with a clock that is 2 minutes fast still shows the accurate countdown
- [ ] A participant with a clock that is 2 minutes slow still shows the accurate countdown
- [ ] Can click the timer icon, click add time, it works

**Comment by jordanh** [comment]
Test cycle 1 complete!

File: src/server/graphql/mutations/helpers/createNewMeetingPhases.ts [hunk 1]
```diff
@@ -29,6 +29,32 @@ const primePhases = (phases: GenericMeetingPhase[]) => {
...
+  return r
+    .table('NewMeeting')
+    .getAll(teamId, {index: 'teamId'})
+    .filter({isLegacy: false}, {default: true})
+    // .orderBy(r.desc('endedAt'))
+    .concatMap((row) => row('phases'))
+    .concatMap((row) => row('stages'))
+    .filter((row) => row.hasFields('startAt', 'endAt'))
+    // convert seconds to ms
+    .merge((row) => ({
```
Comment (jordanh) [inline src/server/graphql/mutations/helpers/createNewMeetingPhases.ts]: This is a great query

Comment (jordanh) [reply src/server/graphql/mutations/helpers/createNewMeetingPhases.ts]: It makes me wonder if we shouldn't be emitting segment events per stage anyway so we could easily interrogate this stuff in amplitude (vs. needing to poke the database)

File: src/server/database/types/GenericMeetingStage.ts [hunk 2]
```diff
@@ -1,13 +1,48 @@
 import shortid from 'shortid'
+import ms from 'ms'
+
+const MAX_SYNC_STAGE_DURATION = ms('1h')
+
+// https://stackoverflow.com/a/20811670/3155110
+const filterOutliers = (someArray: number[]) => {
```
Comment (jordanh) [inline src/server/database/types/GenericMeetingStage.ts]: This is 🔥

File: src/server/database/types/GenericMeetingStage.ts [hunk 3]
```diff
@@ -1,13 +1,48 @@
...
   isNavigableByFacilitator = false
   startAt: Date | undefined = undefined
   endAt: Date | undefined = undefined
+  scheduledEndTime: Date | undefined | null
+  suggestedEndTime: Date | undefined
+  suggestedTimeLimit: number | undefined
   viewCount = 0
-  constructor (public phaseType: string) {}
+
+  constructor (public phaseType: string, durations: number[] | undefined) {
```
Comment (jordanh) [inline src/server/database/types/GenericMeetingStage.ts]: `phaseType` doesn't appear to be used here. Are you thinking you'd use it in the future?

File: src/server/database/types/GenericMeetingStage.ts [hunk 4]
```diff
@@ -1,13 +1,48 @@
 import shortid from 'shortid'
+import ms from 'ms'
+
+const MAX_SYNC_STAGE_DURATION = ms('1h')
```
Comment (jordanh) [inline src/server/database/types/GenericMeetingStage.ts]: Reasonable.

File: src/server/graphql/mutations/setStageTimer.ts [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    }
+
+    const stageRes = findStageById(phases, facilitatorStageId)!
+    const {stage} = stageRes
+    if (scheduledEndTime) {
+      if (timeRemaining) {
+        stage.isAsync = false
+        const actualTimeRemaining = scheduledEndTime - now.getTime()
+        const badClientClock = Math.abs(timeRemaining - actualTimeRemaining) > BAD_CLOCK_THRESH
+        stage.scheduledEndTime = badClientClock ? now + timeRemaining - AVG_PING : scheduledEndTime
```
Comment (jordanh) [inline src/server/graphql/mutations/setStageTimer.ts]: whoa.

File: src/server/graphql/mutations/setStageTimer.ts [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    await r
+      .table('NewMeeting')
+      .get(meetingId)
+      .update({
+        phases,
+        updatedAt: now
+      })
+
+    const data = {meetingId, stageId: facilitatorStageId}
+    publish(TEAM, teamId, SetStageTimerPayload, data, subOptions)
```
Comment (jordanh) [inline src/server/graphql/mutations/setStageTimer.ts]: It's so easy to tell what's persisted and what's just published here after it's been validated/tweaked. I love it. These are boss-mode patterns.

File: src/universal/components/StageTimerHourPicker.tsx [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+  menuProps: MenuProps
+  onClick: (n: number) => void
+}
+
+const options = [...Array(48).keys()].map((n) => n * ms('30m'))
+
+const StageTimerHourPicker = (props: Props) => {
+  const {menuProps, endTime, onClick} = props
+  const currentValue = endTime.getHours() * ms('1h') + endTime.getMinutes() * ms('1m')
+  console.log('curVal', currentValue)
```
Comment (jordanh) [inline src/universal/components/StageTimerHourPicker.tsx]: remove me

File: src/universal/utils/relativeDate.ts [hunk 8]
```diff
@@ -18,6 +19,33 @@ interface Opts {
   smallDiff?: string
 }
 
+export const countdown = (date: string | Date) => {
```
Comment (jordanh) [inline src/universal/utils/relativeDate.ts]: How many times do you think a function like this has been written by folks?

**Comment by mattkrick** [comment]
remaining issues addressed in add-timer-notifications

**Comment by mattkrick** [comment]
merging to master, we can handle fixups on the next branch

**Pull request merged**